apps/choleskey/choleskey_stdpar_snd.cpp

```diff
@@ -0,0 +1,210 @@
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+//
+// This example provides a stdexec(senders/receivers) implementation for choleskey decomposition code.
+#include <algorithm>
+#include <exec/any_sender_of.hpp>
+#include <experimental/mdspan>
+#include <iostream>
+#include <numeric>
+#include <stdexec/execution.hpp>
+#include <vector>
+#include "argparse/argparse.hpp"
+#include "commons.hpp"
+#include "exec/static_thread_pool.hpp"
+
+#include "matrixutil.hpp"
+// using namespace stdexec;
+
+using namespace std;
+
+struct solver {
+
+  using view_2d = std::extents<int, std::dynamic_extent, std::dynamic_extent>;
+
+  template <typename T>
+  std::vector<std::vector<T>> Cholesky_Decomposition(std::vector<T>& vec, int n,
+                                                     int np) {
+
+    // test here first, scheduler from a thread pool
+    exec::static_thread_pool pool(np);
+    stdexec::scheduler auto sch = pool.get_scheduler();
+    stdexec::sender auto begin = stdexec::schedule(sch);
+
+    std::vector<std::vector<T>> lower(n, std::vector<T>(n, 0));
+
+    auto matrix_ms =
+        std::mdspan<T, view_2d, std::layout_right>(vec.data(), n, n);
+
+    auto multiplier_lambda = [=](auto a, auto b) {
+      return a * b;
+    };
+
+    for (int i = 0; i < matrix_ms.extent(0); i++) {
+      for (int j = 0; j <= i; j++) {
+        // avoid over parallelize
+        if (j == 0) {
+          np = 1;
+        } else if (j > 0 && np > j) {
+          np = j;
+        }
+
+        if (j == i)  // summation for diagonals
+        {
+
+          if (i == 0 && j == 0) {
+            lower[j][j] = std::sqrt(matrix_ms(i, j));
+          } else {
+
+            std::vector<T> sum_vec(np);  // sub res for each piece
+            int size = j;  // there are j elements need to be calculated(power)
+
+            stdexec::sender auto send1 =
+                stdexec::bulk(begin, np,
+                              [&](int piece) {
+                                int start = piece * size / np;
+                                int chunk_size = size / np;
+                                int remaining = size % np;
+                                chunk_size += (piece == np - 1) ? remaining : 0;
+
+                                sum_vec[piece] = std::transform_reduce(
+                                    std::execution::par,
+                                    counting_iterator(start),
```

Comment:
This is not correct since `counting_iterator(start)` and `counting_iterator(start +N)` are two separate objects and hence not iterable

Reply:
Also being caught by the compiler. see below:

<img width="1528" alt="image" src="https://github.com/mhaseeb123/nvstdpar/assets/14217455/7e2a8ac2-39bc-47b7-9150-eefb7480e6bc">

Reply:
Nevermind, this error was coming from `libcudart/11.7` which is already loaded into modules and takes precedence even with nvhpc/23.7 compiler. Doing a `ml unload cudatoolkit` and rerunning cmake and make uses the latest `cudart/12.x` from `nvhpc/23.x` module and works fine.

Reply:
great! thanks a lot. 

I have not encountered build issue. 
I  build with following options:
`cmake .. -DCMAKE_CXX_COMPILER=$(which nvc++) -DCMAKE_C_COMPILER=$(which nvc) -DCMAKE_BUILD_TYPE=Release -DSTDPAR=gpu`

modules:
`ml use /global/cfs/cdirs/m1759/wwei/nvhpc_23_7/modulefiles ; ml unload cudatoolkit ; ml nvhpc/23.1 cmake/3.24`